### PR TITLE
[APM2-578] Fetch capabilities

### PR DESCRIPTION
**Pull request by new4762**
#### 1. Objective

Currently the payment options that display in checkout page didn’t fetch from capabilities api 
so this PR focus to update checkout page to display payment options correctly 
for `installment` , `internet banking`  and `mobile banking`

- Update installment to display from fetching capabilities api  
- Update internet banking to display from fetching capabilities api  
- Update mobile banking to display from fetching capabilities api  
- Update mobile banking to use currency options from capabilities api  

This section will be used in the release notes. 

**Related information**:
Related issue(s): #< GitHub ticket number > (optional)

- https://omise.atlassian.net/browse/APM2-690
- https://omise.atlassian.net/browse/APM2-691
- https://omise.atlassian.net/browse/APM2-692

#### 2. Description of change

Replace statics display payments options to be dynamic display from capabilities api .
effect to `installment` , `internet banking`  and `mobile banking`

 
#### 3. Quality assurance

Test payments option displays correctly and  can created charge to staging 

**🔧 Environments:**

- **Platform version**: Magento CE 2.2.3
- **Omise plugin version**: Omise-Magento 2.23.1
- **Omise plugin version**: Omise-php v2.13
- **PHP version**: 7.1

**✏️ Details:**

Try switching `source_type` config to check is really dynamic display and testing create charges to make sure it correct sources type
 
#### 4. Impact of the change

`installment` , `internet banking`  and `mobile banking` will display only enabled sources type

**Internet banking**
image
image

**Installment**
image
image
image

**Mobile banking**
image
image

Note: Please provide a screenshot if your changed impact to UI.

#### 5. Priority of change

Normal

#### 6. Additional Notes

Any further information that you would like to add.

### CODE REVIEW

File: view/frontend/web/js/view/payment/method-renderer/omise-offsite-installment-method.js [hunk 1]
```diff
@@ -165,6 +234,32 @@ define(
                 );
             },
 
+            /**
+             * Get installment term ko.observable by name
+             *
+             * @return {string|null}
+             */
+            getObservableTrem: function (name) {
```
Comment (tookaa2) [inline view/frontend/web/js/view/payment/method-renderer/omise-offsite-installment-method.js:242]: should this be getObservableTerm?

File: view/frontend/web/template/payment/offsite-installment-form.html [hunk 2]
```diff
@@ -43,231 +45,41 @@
...
-
+                        <select style="margin-top:10px" data-bind="options: $parent.getInstallmentTerms(logo),
+                                                      optionsText: function (item) {
+                                                          return item.label
+                                                      },
+                                                      optionsValue: function (item) {
+                                                        return item.key
+                                                      },
+                                                      valueAllowUnset:true,
+                                                      value: $parent.getObservableTrem(obs),
```
Comment (tookaa2) [inline view/frontend/web/template/payment/offsite-installment-form.html:79]: this too

File: view/frontend/web/js/view/payment/method-renderer/omise-offsite-installment-method.js (hunk 1 above)
Comment (new4762) [reply view/frontend/web/js/view/payment/method-renderer/omise-offsite-installment-method.js]: nice catch !!
Fixed -> https://github.com/omise/omise-magento/pull/337/commits/72d9606215fd4c67893798f6857f0487081c6435 🙏

**Review by tookaa2 — APPROVED** [review]

**Review by hpariwat — APPROVED** [review]

File: Model/Ui/CapabilitiesConfigProvider.php [hunk 3]
```diff
@@ -47,6 +48,12 @@ public function getConfig()
                     $backendsFpx = $this->capabilities->getBackendsByType(Fpx::TYPE);
                     $configs['fpx']['banks'] = $backendsFpx ? current($backendsFpx)->banks : [];
                     break;
+                case Mobilebanking::CODE:
+                    $configs['mobile_banking'] = $this->capabilities->retrieveMobileBankingBackends();
+                    break;
+                case Internetbanking::CODE:
+                    $configs['internet_banking'] = $this->capabilities->getBackendsByType("internet_banking");
```
Comment (kan98) [inline Model/Ui/CapabilitiesConfigProvider.php:55]: why don't we reuse the function `getBackendsByType`  for mobile banking too? just wondering if there was a reason

File: view/frontend/web/js/view/payment/method-renderer/omise-offsite-mobilebanking-method.js [hunk 4]
```diff
@@ -55,6 +95,26 @@ define(
...
+            *
+            * @return {Array}
+            */
+            get_available_providers: function () {
+                let _providers = Object.values(window.checkoutConfig.mobile_banking);
+
+                return this.providers.filter((a1) => _providers.find(a2 => {
+                    if (a1.id === a2._id) {
+                        // set currencies from api if is undefined use default value
+                        if (a2?.currencies !== undefined) {
```
Comment (kan98) [inline view/frontend/web/js/view/payment/method-renderer/omise-offsite-mobilebanking-method.js:110]: might be better to use `null` here as null covers both null and undefined cases in javascript

File: view/frontend/web/template/payment/offsite-installment-form.html [hunk 5]
```diff
@@ -289,4 +101,4 @@
             </div>
         </div>
     </div>
-</div>
+</div>
```
Comment (kan98) [inline view/frontend/web/template/payment/offsite-installment-form.html:104]: EOF

File: Model/Ui/CapabilitiesConfigProvider.php (hunk 3 above)
Comment (new4762) [reply Model/Ui/CapabilitiesConfigProvider.php:55]: Is bc our omsie-php is using `aegaeon` version for capabilities and we set is like this
``` ruby
    "internet_banking_bay" => "internet_banking",
    "internet_banking_ktb" => "internet_banking",
    "internet_banking_scb" => "internet_banking",
    "internet_banking_bbl" => "internet_banking",
    "alipay" => "alipay",
    "installment_bay" => "installment",
    "installment_kbank" => "installment",
    "installment_ktc" => "installment",
    "installment_bbl" => "installment",
    "installment_first_choice" => "installment",
    "installment_citi" => "installment",
    "installment_scb" => "installment",
    "installment_ttb" => "installment",
    "installment_uob" => "installment",
    "installment_ezypay" => "installment",
    "mobile_bป--anking_scb" => "mobile_banking_scb",
    "mobile_banking_kbank" => "mobile_banking_kbank",
    "mobile_banking_bbl" => "mobile_banking_bbl",
    "mobile_banking_bay" => "mobile_banking_bay",
    "mobile_banking_ocbc_pao" => "mobile_banking_ocbc_pao",
   ```

File: view/frontend/web/js/view/payment/method-renderer/omise-offsite-mobilebanking-method.js (hunk 4 above)
Comment (new4762) [reply view/frontend/web/js/view/payment/method-renderer/omise-offsite-mobilebanking-method.js]: Noted 📓  -> https://github.com/omise/omise-magento/pull/337/commits/a0dde384bd84e631ba32d8c03502111eb83c8f84

File: view/frontend/web/template/payment/offsite-installment-form.html (hunk 5 above)
Comment (new4762) [reply view/frontend/web/template/payment/offsite-installment-form.html]: Nice catch! Fixed -> https://github.com/omise/omise-magento/pull/337/commits/a0dde384bd84e631ba32d8c03502111eb83c8f84

File: Model/Ui/CapabilitiesConfigProvider.php (hunk 3 above)
Comment (kan98) [reply Model/Ui/CapabilitiesConfigProvider.php:55]: ohhh I see, that makes sense now thanks for explaining 🙂

**Review by kan98 — APPROVED** [review]

**Comment by som-m** [comment]
All good. But I'm just wondering if we should just hide the payment method if there is no option. Please see mobile banking section below.

image
image

**Comment by new4762** [comment]
> All good. But I'm just wondering if we should just hide the payment method if there is no option. Please see mobile banking section below.
> 
> image image

This one is will be another tasks krub 
bc this one is need to investigate and apply to payment options on magento and woocommerce krub

**Review by som-m — APPROVED** [review]

**Review by aashishgurung — APPROVED** [review]

**Pull request merged**